homeassistant/components/xiaomi_miio/fan.py

```diff
@@ -1087,16 +1099,22 @@ async def async_set_percentage(self, percentage: int) -> None:
             await self.async_turn_off()
             return
 
-        await self._try_command(
-            "Setting fan speed percentage of the miio device failed.",
-            self._device.set_speed,
-            percentage,
+        speed = math.ceil(
+            percentage_to_ranged_value((1, self._speed_count), percentage)
         )
-        self._percentage = percentage
 
+        # if the fan is not on, we have to turn it on first
         if not self.is_on:
             await self.async_turn_on()
-        else:
+
+        result = await self._try_command(
+            "Setting fan speed percentage of the miio device failed.",
+            self._device.set_speed,
+            speed,
+        )
+
+        if result:
+            self._percentage = ranged_value_to_percentage((1, self._speed_count), speed)
```

Comment:
@MartinHjelmare I was thinking to set _percentage directly to the ranged value which would make it snap to the range borders instead of doing the same only when we get the state update in `_handle_coordinator_update directly` from the device. What do you think? 

Behavior before: 
 * set to percentage, e.g. 45%
 * will stay at 45% until we get the update of the device that speed 2 is now selected which will update percentage to something like 33% due to the ranged nature of mapping from speed to %

New behavior: 
 * set percentage, e.g. 45%
 * as soon as the command is executed the percentage snaps directly to the correctly ranged percentage, i.e. 33%.

The new behavior would omit the seemingly random jumps of the percentage due to the update from the device a few seconds after setting a new value.

Reply:
Sounds good.